experimenter/experimenter/nimbus-ui/src/components/PageEditMetrics/FormMetrics/index.tsx

```diff
@@ -128,6 +128,69 @@ const FormMetrics = ({
     [isLoading, onSave, handleSubmit, primaryOutcomes, secondaryOutcomes],
   );
 
+  // Two new variables to handle the primary and secondary states.
+  const [primaryValid, setPrimaryValid] = useState(false);
+  const [secondaryValid, setSecondaryValid] = useState(false);
+
+  useEffect(() => {
+    const primaryContainerDiv = document.querySelector(
+      '[data-testid="primary-outcomes"] .css-g1d714-ValueContainer',
+    );
+    const secondaryContainerDiv = document.querySelector(
+      '[data-testid="secondary-outcomes"] .css-g1d714-ValueContainer',
+    );
+    const primarySelectContainer = document.querySelector(
+      '[data-testid="primary-outcomes"] .css-yk16xz-control',
+    );
+    const secondarySelectContainer = document.querySelector(
+      '[data-testid="secondary-outcomes"] .css-yk16xz-control',
+    );
+
+    if (primaryContainerDiv) {
+      if (primaryValid) {
+        primaryContainerDiv.classList.add(
+          "form-control",
+          "is-valid",
+          "valid-right",
+        );
+        primarySelectContainer?.classList.add("border-success");
+      } else {
+        primaryContainerDiv.classList.remove(
+          "form-control",
+          "is-valid",
+          "valid-right",
+        );
+        primarySelectContainer?.classList.remove("border-success");
+      }
+    }
+
+    if (secondaryContainerDiv) {
+      if (secondaryValid) {
+        secondaryContainerDiv.classList.add(
+          "form-control",
+          "is-valid",
+          "valid-right",
+        );
+        secondarySelectContainer?.classList.add("border-success");
+      } else {
+        secondaryContainerDiv.classList.remove(
+          "form-control",
+          "is-valid",
+          "valid-right",
+        );
+        secondarySelectContainer?.classList.remove("border-success");
+      }
+    }
+  }, [primaryValid, secondaryValid]);
+
+  useEffect(() => {
+    setPrimaryValid(primaryOutcomes.length > 0);
+  }, [primaryOutcomes]);
+
+  useEffect(() => {
+    setSecondaryValid(secondaryOutcomes.length > 0);
+  }, [secondaryOutcomes]);
```

Comment:
few things 
Here are some optimizations you can make to this code:
-  Use useRef instead of querySelector:
    querySelector is not recommended for performance reasons because it searches the entire DOM tree. Instead, you can use useRef to reference the specific DOM element you want to manipulate.
- Combine the two useEffect hooks into one:
    Since both useEffect hooks are checking for changes in the same state variables, you can combine them into one hook to reduce code duplication.
  -Use a single state variable instead of two:
    Since the only difference between primaryValid and secondaryValid is the DOM elements they reference, you can use a single state variable to keep track of both. You can also combine the two setPrimaryValid and setSecondaryValid into one function that takes an argument to determine which one to update.
- Avoid unnecessary classList operations:
    If a DOM element already has a certain class, there's no need to add or remove it again. You can use classList.toggle() to add or remove a class based on whether it's already present.

you can do something like this probably, I haven't tested the code its a suggestion
```javascript 
const [valid, setValid] = useState({ primary: false, secondary: false });

const primaryContainerDivRef = useRef();
const secondaryContainerDivRef = useRef();
const primarySelectContainerRef = useRef();
const secondarySelectContainerRef = useRef();

useEffect(() => {
  const primaryContainerDiv = primaryContainerDivRef.current;
  const secondaryContainerDiv = secondaryContainerDivRef.current;
  const primarySelectContainer = primarySelectContainerRef.current;
  const secondarySelectContainer = secondarySelectContainerRef.current;

  if (primaryContainerDiv && primarySelectContainer) {
    primaryContainerDiv.classList.toggle("form-control", valid.primary);
    primaryContainerDiv.classList.toggle("is-valid", valid.primary);
    primaryContainerDiv.classList.toggle("valid-right", valid.primary);
    primarySelectContainer.classList.toggle("border-success", valid.primary);
  }

  if (secondaryContainerDiv && secondarySelectContainer) {
    secondaryContainerDiv.classList.toggle("form-control", valid.secondary);
    secondaryContainerDiv.classList.toggle("is-valid", valid.secondary);
    secondaryContainerDiv.classList.toggle("valid-right", valid.secondary);
    secondarySelectContainer.classList.toggle("border-success", valid.secondary);
  }
}, [valid]);

useEffect(() => {
  setValid((prevState) => ({
    ...prevState,
    primary: primaryOutcomes.length > 0,
  }));
}, [primaryOutcomes]);

useEffect(() => {
  setValid((prevState) => ({
    ...prevState,
    secondary: secondaryOutcomes.length > 0,
  }));
}, [secondaryOutcomes]);

Reply:
@yashikakhurana Thank you for the feedback. I will work on the changes asap.

Reply:
@yashikakhurana when I placed a ref on the `Select` component. I got an error that it was not compatible with the type expected by the `ref` prop of the `Select` component.

So I placed the ref on the `Form.Group` components. And grabbed the necessary classLists to be updated using `getElementsByClassName`[0]. This is because using `getElementByClassName` instead of `querySelector` can help with performance since it restricts the search to a specific set of elements with the specified class name.

It also allowed me to limit the refs to just two for each `Form.Group`.